src/relay/pass/partial_eval.cc

```diff
@@ -232,12 +233,145 @@ struct SFuncNode : StaticNode {
   TVM_DECLARE_NODE_TYPE_INFO(SFuncNode, StaticNode);
 };
 
-RELAY_DEFINE_NODE_REF(SFunc, SFuncNode, Value);
+RELAY_DEFINE_NODE_REF(SFunc, SFuncNode, Static);
 
 Static MkSFunc(const Func& func) {
   return Static(make_node<SFuncNode>(func));
 }
 
+
+class FuelNode;
+/*! \brief A meet-semilattice with finite descending chain.
+ * It mean that we can meet two element to get an element,
+ * and for every element, there is only finite amount of meet before getting back the same element.
+ *
+ * Every time we recurse, we do a meet and require that progress must be made.
+ * This make sure we do not recurse infinitely in the Partial Evaluator.
+ */
+class Fuel : public NodeRef {
+ public:
+  Fuel() {}
+  explicit Fuel(NodePtr<Node> n) : NodeRef(n) {}
+  const FuelNode* operator->() const;
+
+  using ContainerType = FuelNode;
+};
+
+class FuelNode : public RelayNode {
+ public:
+  // Please implement one of the following function or there will be infinite loop.
+  /*! \brief return the new Fuel, and whether progress is made.
+   *
+   * Note that progress is not symmetric - it only measure progress for (*this).
+   *
+   * Thus, if the generated is smaller then the argument of Meet,
+   * and the generated is not smaller then (*this),
+   * progress should be false.
+   */
+  virtual std::tuple<Fuel, bool> Meet(const Fuel& f) const {
+    bool progress = false;
+    auto ret = Meet(f, &progress);
+    return std::make_tuple(ret, progress);
+  }
+  /*! \brief return the new Fuel, and write true only iff progress is made. */
```

Comment:
```suggestion
  /*! \brief return the new Fuel, and write true iff progress is made. */
```

Reply:
I am trying to communicate that if progress is already true, in no case it will be written false. I will rewordsmith.